web/publicendpointresource.go

```diff
@@ -291,118 +291,35 @@ func restAddPort(w *rest.ResponseWriter, r *rest.Request, client *node.ControlCl
-	if err != nil {
-		err := fmt.Errorf("Port must be a number between 1 and 65536")
-		glog.Error(err)
-		restServerError(w, err)
-		return
-	}
-
-	if port < 1 || port > 65535 {
-		err := fmt.Errorf("Port must be a number between 1 and 65536")
-		glog.Error(err)
-		restServerError(w, err)
-		return
-	}
-
-	switch port {
-	case 5000:
-		fallthrough
-	case 8080:
-		err := fmt.Errorf("Port %d is reserved", port)
-		glog.Error(err)
-		restServerError(w, err)
-		return
-	}
-
-	var services []svc.Service
-	var serviceRequest dao.ServiceRequest
-	if err := client.GetServices(serviceRequest, &services); err != nil {
-		err := fmt.Errorf("Could not get services: %v", err)
-		glog.Error(err)
-		restServerError(w, err)
-		return
-	}
-
-	var service *svc.Service
-	for _, _service := range services {
-		if _service.ID == request.ServiceID {
-			service = &_service
-			break
-		}
-	}
 
-	if service == nil {
-		err := fmt.Errorf("Could not find service: %s", request.ServiceID)
-		glog.Error(err)
-		restServerError(w, err)
-		return
-	}
+	facade := ctx.getFacade()
+	dataCtx := ctx.getDatastoreContext()
 
-	//checkout other ports for redundancy
-	for _, service := range services {
-		if service.Endpoints == nil {
-			continue
-		}
-
-		for _, endpoint := range service.Endpoints {
-			for _, epPort := range endpoint.PortList {
-				if scrubbedPort == epPort.PortAddr {
-					err := fmt.Errorf("Port %s already defined for service: %s", epPort.PortAddr, service.Name)
-					glog.Error(err)
-					restServerError(w, err)
-					return
-				}
-			}
-		}
-	}
-
-	// Check to make sure the port is available.  Don't allow adding a port if it's already being used.
-	err = checkPort("tcp", fmt.Sprintf("%s", scrubbedPort))
+	// Get the service for this service id.
+	service, err := facade.GetService(dataCtx, request.ServiceID)
 	if err != nil {
-		glog.Error(err)
+		err = fmt.Errorf("Could not find service %s: %s", request.ServiceID, err)
 		restServerError(w, err)
 		return
 	}
 
-	err = service.AddPort(request.Application, scrubbedPort, request.UseTLS, request.Protocol)
+	port, err := facade.AddPublicEndpointPort(dataCtx, request.ServiceID, request.Application,
+		request.PortName, request.UseTLS, request.Protocol, true)
 	if err != nil {
 		glog.Errorf("Error adding port to service (%s): %v", service.Name, err)
 		restServerError(w, err)
 		return
 	}
 
 	glog.V(2).Infof("Port (%s) added to service (%s), UseTLS=%s, protocol=%s",
-					request.PortName, service.Name, request.UseTLS, request.Protocol)
-
-	var unused int
-	err = client.UpdateService(*service, &unused)
-	if err != nil {
-		glog.Errorf("Unexpected error adding port to service (%s): %v", service.Name, err)
-		restServerError(w, err)
-		return
-	}
-
-	glog.V(2).Infof("Service (%s) updated", service.Name)
+		port.PortAddr, service.Name, port.UseTLS, port.Protocol)
 
 	// Restart the service if it is running
 	if service.DesiredState == int(svc.SVCRun) || service.DesiredState == int(svc.SVCRestart) {
```

Comment:
I would move this restart logic inside Facade.AddPort() so that the CLI add-port and the UI add-port have the same results.  If a use-case emerges for adding the port without restarting the service, we can add extra parameters to the AddPort() method to control that.